enhancements/windows-containers/monitoring-windows-nodes.md

```diff
@@ -0,0 +1,182 @@
+  do not get same metrics from cAdvisor for Windows nodes. This becomes a
+  hindrance to display pod graphs by creating custom recording rules to use same
+  console queries as Linux workloads. To mitigate this, use metrics exposed by
+  the windows_exporter to display pod graphs as mentioned in the [Future
+   Plans](#future-plans)
+  is required. This also requires changes in console queries that support OS
+  specific metrics.
+
+## Design Details
+
+As we are not able to run windows_exporter as a [container](#justification)
+on the Windows Node, to capture data from windows_exporter, WMCO creates a
+`windows-machine-config-operator-metrics` Service without selectors and
+manually defines Endpoints object for that service. The Endpoints object has
+entries for the endpoints `<internal-ip>:9182/metrics`, exposed by
+windows_exporter for every Windows node. Once the Service and Endpoints
+object is created, WMCO ensures that a Service Monitor for `windows-machine-config-operator-metrics`
+Service is running so that the Prometheus operator can discover the targets
+created above to scrape Windows metrics. Following design details reflect the
+current approach and future plans to enable monitoring support for Windows.
+
+### Current State
+
+To enable basic monitoring support for Windows node, WMCO has done the
+following:
+
+* Build and add windows_exporter binary to WMCO payload.
+* Install windows_exporter on the Windows nodes and ensuring
+  that it runs as a Windows service.
+* Add `openshift.io/cluster-monitoring=true` label to the
+ `openshift-windows-machine-config-operator` namespace so that cluster
+  monitoring stack will pick up the Service Monitor created by WMCO.
+* Add privileges to WMCO to create Services, Endpoints, Service Monitor in
+  the `openshift-windows-machine-config-operator` namespace.
+* Create a Service and Endpoints object in `openshift-windows-machine-config
+-operator` namespace that point to windows_exporter endpoint. WMCO uses default
+  values to define metrics endpoint, `<internal-ip>:9182/metrics`,
+  exposed by windows_exporter for every Windows node. The Endpoints object
+  created in the namespace consist of subsets of endpoints from all the
+  Windows nodes.
+* Create a Service Monitor in `openshift-windows-machine-config-operator`
+  namespace for Service created above.
+  
+To display node graphs WMCO has done the following:
+
+* Add custom Prometheus rules in `openshift-windows-machine-config-operator`
+  namespace. The custom recording rules are created using Windows metrics
+  exposed by the windows_exporter and have the same names as Linux
+  recording rules. This is to make use of same console queries as Linux.
+* Note that WMCO is unable to display pod graphs for the Windows Nodes
+  with the current implementation. See [Risks and Mitigations](#risks-and-mitigations)
+  for details.
+  
+### Future Plans
+
+As we move forward, our plan to display monitoring graphs is to create a [common
```

Comment:
I'm not sure we want to offload the recording rules to the cluster-monitoring-operator (CMO) repository. My thinking was more that the Linux-specific rules would live in CMO while the Windows-specific rules would live in WMCO. Otherwise I fear that ownership will be diluted. 
Another option would be to move all recording rules to the console operator but again my intuition is that the rules should stay close to the things that expose the metrics.

Reply:
I am not sure how distributed ownership of recording rules will work if we are thinking of creating a unified interface.  My opinion on this is that WMCO will be responsible for making sure that we receive metrics from windows nodes and CMO handles the complexity of grouping these(Linux + Windows) metrics to make the queries platform-independent. This will help in reducing the complexity on the Console side as well. [See](https://issues.redhat.com/browse/WINC-530)

Reply:
The issue with pushing down all recording rules to CMO is that CMO doesn't have a way to validate that Windows-specific rules are correct. We should probably start by listing all metrics that are used by the console and identifying the gaps from the Windows side. 

Also what is the general testing strategy to ensure that the console works properly for Windows?
@spadgett, I think that there are e2e console tests validating that the dashboards display correct result?

Reply:
>We should probably start by listing all metrics that are used by the console and identifying the gaps from the Windows side.

@VaishnaviHire please look into this.

>Also what is the general testing strategy to ensure that the console works properly for Windows?

We were discussing this as a console change broke the filesystem graphs for Windows nodes: [BZ 1930347](https://bugzilla.redhat.com/show_bug.cgi?id=1930347). The approach we were thinking about was to add a Windows e2e to the console repo or add console tests for validating the dashboards to the WMCO repo depending on whatever is easier. We will defer to @spadgett on what the better approach is.

Reply:
@simonpasquier  I have included a list of metrics used for displaying console graphs and the corresponding Windows metrics in this [doc](https://docs.google.com/document/d/1-So18TTztD3ac3S9_8PigVkWcbatJC3IHAU5dOeWljc/edit?usp=sharing). Please take a look.  I have also added the cases where the existing queries won`t work for Windows.

Reply:
Updated to reflect `Windows-specific` rules will be added in WMCO.

Reply:
@simonpasquier Updated the `future plans` section to reflect the discussion from our meeting. Please review.